### PR TITLE
GG-30086 - remove empty locked pages from log

### PR DESCRIPTION
**Pull request by Wuchte**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/PageLockDump.java [hunk 1]
```diff
@@ -16,17 +16,50 @@
...
-    /**
-     * @param dumpProcessort Apply dump processor.
-     */
-    void apply(DumpProcessor dumpProcessort);
-
-    /**
-     * @return Dump creation time.
-     */
-    long time();
+public abstract class PageLockDump {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/PageLockDump.java:24]: Can we make it package-private?
Can we make most of these classes package-private?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java [hunk 2]
```diff
@@ -47,291 +46,290 @@
...
-     * @return String representation of dump.
-     */
-    public static String toStringDump(PageLockDump pageLockDump) {
-        StringBuilder sb = new StringBuilder();
+/** */
+class ToStringDumpProcessor {
 
-        ToStringHelper proc = new ToStringHelper(sb, strucutreIdMapFunc(pageLockDump));
+    /** */
+    private final Function<Integer, String> strucutreIdMapFunc;
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java:53]: java.util.function.IntFunction

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java [hunk 3]
```diff
@@ -47,291 +46,290 @@
...
+        this.sb = sb;
+        this.strucutreIdMapFunc = strucutreIdMapFunc;
     }
 
-    /** */
-    private static Function<Integer, String> strucutreIdMapFunc(PageLockDump pageLockDump) {
-        if (pageLockDump instanceof ThreadPageLocksDumpLock) {
-            ThreadPageLocksDumpLock dump = (ThreadPageLocksDumpLock)pageLockDump;
+    /** Helper class for track lock/unlock count. */
+    class LockState {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java:65]: Make it static

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java [hunk 4]
```diff
@@ -47,291 +46,290 @@
...
-                    ", nextOpStructureId=" + strucutreIdMapFunc.apply(nextOpStructureId)
-                    + " [pageIdHex=" + hexLong(nextOpPageId)
-                    + ", partId=" + partId(nextOpPageId) + ", pageIdx=" + pageIndex(nextOpPageId)
-                    + ", flags=" + hexInt(flag(nextOpPageId)) + "]" + U.nl());
-            }
+            appendThreadInfo(sb, threadState);
+        }
 
-            sb.append(lockedPagesInfo(holdetLocks)).append(U.nl());
+        sb.append(lockedPagesInfo(holdetLocks)).append(U.nl());
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java:215]: What if threadState == null and holdetLocks is empty? Will you print it?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java (hunk 2 above)
Comment (Wuchte) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java:53]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java (hunk 3 above)
Comment (Wuchte) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpProcessor.java:65]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java:2]: Still 2020 :)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        Thread asyncLockUnlock = new Thread(() -> {
+            pageLockTracker.onReadLock(4, 32, 1, 64);
+        }, "async-lock-unlock");
+        asyncLockUnlock.start();
+        asyncLockUnlock.join();
+        long threadIdInLog = asyncLockUnlock.getId();
+
+        SharedPageLockTrackerDump pageLockDump = pageLockTracker.dump();
+
+        // hack to have same timestamp in test
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java:51]: Start it with capital letter and end with a dot.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        Field timeField = PageLockDump.class.getDeclaredField("time");
+        timeField.setAccessible(true);
+
+        for (ThreadPageLockState state : pageLockDump.threadPageLockStates) {
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java:55]: And remove braces.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            timeField.setLong(state.pageLockDump, 1596173397167L);
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java:56]: There's a GridTestUtils#setFieldValue method.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        String dumpStr = ToStringDumpHelper.toStringDump(pageLockDump);
+
+        System.out.println("Dump saved:" + dumpStr);
+
+        String expectedLog = "Page locks dump:" +
+            U.nl() +
+            U.nl() +
+            "Thread=[name=async-lock-unlock, id=" + threadIdInLog + "], state=TERMINATED" + U.nl() +
+            "Locked pages = [32[0000000000000020](r=1|w=0)]\n" +
+            "Locked pages log: name=async-lock-unlock time=(1596173397167, 2020-07-31 08:29:57.167)" + U.nl() +
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java:70]: Will it pass on TC? What timezone are we using here?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+        for (ThreadPageLockState state : pageLockDump.threadPageLockStates) {
+            timeField.setLong(state.pageLockDump, 1596173397167L);
+        }
+
+        assertNotNull(pageLockDump);
+
+        String dumpStr = ToStringDumpHelper.toStringDump(pageLockDump);
+
+        System.out.println("Dump saved:" + dumpStr);
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/diagnostic/pagelocktracker/dumpprocessors/ToStringDumpHelperTest.java:63]: There's no point in that, string will be dumped if it doesn't match with the expected result.

File: modules/core/src/test/java/org/apache/ignite/testsuites/IgnitePdsMvccTestSuite4.java [hunk 11]
```diff
@@ -62,6 +63,7 @@
         ignoredTests.add(PageLockTrackerManagerTest.class);
         ignoredTests.add(SharedPageLockTrackerTest.class);
         ignoredTests.add(ToFileDumpProcessorTest.class);
+        ignoredTests.add(ToStringDumpHelperTest.class);
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/testsuites/IgnitePdsMvccTestSuite4.java:66]: Ok, I see that you added it to ignored tests somewhere, but where is the actual suit that will run it?

**Pull request closed without merging**